sniffles/petabi_rule_writer/petabi_rule_writer.py

```diff
@@ -185,16 +196,66 @@ def check_pcre_compile(re):
 # Formats background traffic rule to a petabi rule format.
 # There can only be 1 background traffic rule in a whole rule file.
 # Output: foramtted background traffic rule
-def formatBackgroundTrafficRule(background_traffic):
+def formatBackgroundTrafficRule(background_traffic, protocol_dist=None):
     bgTrafficInfo = []
     bgTrafficInfo.append("    <traffic_stream")
     bgTrafficInfo.append("typets=\"BackgroundTraffic\"")
     bgTrafficInfo.append("percentage=\"" + background_traffic + "\"")
+    if protocol_dist:
+        protocol_dict = protocolPercentage(protocol_dist)
+        for protocol in protocol_dict:
+            distribution = protocol + "=\"" + str(protocol_dict[protocol]) + \
+                           "\""
+            bgTrafficInfo.append(distribution)
     bgTrafficInfo.append(">\n")
     background_rule = ' '.join(bgTrafficInfo)
     return background_rule
 
 
+# Label protocol percentage with given distribution.
+# Wildcard (* symbol) is used to denote even percentage among wildcards.
+# Wildcard protocol will not be shown in the rule, but will be applied
+# in sniffles. When all protocols are given a percentage and their sum
+# does not add up to 100, they are treated as ratio.
+# Output: Dictionary containing protocol percentage
+def protocolPercentage(protocol_dist):
+    protocol_percent = OrderedDict()
+    total = 0
+    num_wildcard = 0
+    protocol_list = ['http', 'ftp', 'pop', 'smtp', 'imap']
+    i = 0
+    while i < len(protocol_list):
+        if protocol_dist[i] != "*":
+            protocol_percent[protocol_list[i]] = int(protocol_dist[i])
+        i += 1
+
+    # Change ratio to percentage for protocol if they don't add up to 100
+    if len(protocol_list) == len(protocol_percent):
+        protocol_sum = sumPercentage(protocol_percent)
+        while protocol_sum != 100:
```

Comment:
1. I think we can calculate the protocol sum above.
2. If the sum is too small, just add the remainder to the first entry in the list, or redistribute the remaining evenly amongst all others, or, perhaps better, Error and warn the user.  Silently adjusting might cause errors the user does not expect.
3. Same if it is too big.